File: b2ou/markdown.py
"""
Pure Markdown transformation functions — no I/O, no config dependencies.

All functions take a string and return a string (or ancillary data).
They rely only on the pre-compiled patterns in ``b2ou.constants``.
"""

from __future__ import annotations

import urllib.parse
from typing import Optional

from b2ou.constants import (
    RE_BEAR_EMBED,
    RE_CLEAN_TITLE,
    RE_HIDE_TAGS,
    RE_HTML_IMG_ALT,
    RE_HTML_IMG_SRC,
    RE_HTML_IMG_TAG,
    RE_MD_HEADING,
    RE_MD_IMAGE,
    RE_REF_CLEAN,
    RE_REF_DEF,
    RE_REF_IMP,
    RE_REF_IMG,
    RE_REF_LINK,
    RE_REF_LINK_IMP,
    RE_TAG_PATTERN1,
    RE_TAG_PATTERN2,
    RE_TRAILING_DASH,
)


# ---------------------------------------------------------------------------
# Title sanitisation
# ---------------------------------------------------------------------------

def clean_title(title: str) -> str:
    """Return a filesystem-safe version of *title*.

    The result is capped so that the UTF-8 byte length stays under 240,
    leaving room for a file extension (e.g. ``.md``, ``.textbundle``).
    """
    # Cap characters first (fast path for ASCII-only titles)
    title = title[:225].strip() or "Untitled"
    title = RE_CLEAN_TITLE.sub("-", title)
    title = RE_TRAILING_DASH.sub("", title)
    title = title.strip()

    # Ensure the UTF-8 byte length fits in 240 bytes (255 limit minus extension)
    max_bytes = 240
    encoded = title.encode("utf-8")
    if len(encoded) > max_bytes:
        # Truncate by decoding back with error handling to avoid splitting mid-char
        title = encoded[:max_bytes].decode("utf-8", errors="ignore").rstrip()

    return title


# ---------------------------------------------------------------------------
# Bear → standard Markdown normalisation
# ---------------------------------------------------------------------------

import re as _re

_RE_BEAR_HIGHLIGHT = _re.compile(r'(?<!\:)\:\:(?!\:)(.+?)(?<!\:)\:\:(?!\:)')


def bear_highlight_to_md(text: str) -> str:
    """Convert Bear's ``::highlight::`` syntax to Obsidian's ``==highlight==``."""
    return _RE_BEAR_HIGHLIGHT.sub(r'==\1==', text)


def normalise_bear_markdown(text: str) -> str:
    """Apply all Bear → standard Markdown conversions.

    Currently handles:
    - ``::highlight::`` → ``==highlight==``
    - ``<img>`` HTML tags → ``![alt](src)``
    """
    text = bear_highlight_to_md(text)
    text = html_img_to_markdown(text)
    text = RE_BEAR_EMBED.sub("", text)
    return text


# ---------------------------------------------------------------------------
# Tag handling
# ---------------------------------------------------------------------------

def hide_tags(text: str) -> str:
    """Strip Bear tag lines (lines beginning with #tag) from *text*."""
    return RE_HIDE_TAGS.sub(r"\1", text)


def extract_tags(text: str) -> list[str]:
    """Return all Bear tag strings found in *text* (without the leading #)."""
    tags: list[str] = []
    tags.extend(m[0] for m in RE_TAG_PATTERN1.findall(text))
    tags.extend(m[0] for m in RE_TAG_PATTERN2.findall(text))
    return tags


def sub_path_from_tag(
    base_path: str,
    filename: str,
    text: str,
    make_tag_folders: bool,
    multi_tag_folders: bool,
    only_export_tags: list[str],
    exclude_tags: list[str],
) -> list[str]:
    """
    Return the list of output file paths for a note, based on its tags.

    With *make_tag_folders=False* (default) returns a single path under
    *base_path*.  With *make_tag_folders=True* the note may be written to
    multiple tag-based subdirectories (when *multi_tag_folders=True*).
    """
    import os

    if not make_tag_folders:
        is_excluded = any(("#" + tag) in text for tag in exclude_tags)
        return [] if is_excluded else [os.path.join(base_path, filename)]

    if multi_tag_folders:
        tags = extract_tags(text)
        if not tags:
            return [os.path.join(base_path, filename)]
    else:
        t1 = RE_TAG_PATTERN1.search(text)
        t2 = RE_TAG_PATTERN2.search(text)
        if t1 and t2:
            tag = t1.group(1) if t1.start(1) < t2.start(1) else t2.group(1)
        elif t1:
            tag = t1.group(1)
        elif t2:
            tag = t2.group(1)
        else:
            return [os.path.join(base_path, filename)]
        tags = [tag]

    paths = [os.path.join(base_path, filename)]
    for tag in tags:
        if tag == "/":
            continue
        if only_export_tags:
            if not any(tag.lower().startswith(et.lower())
                       for et in only_export_tags):
                continue
        if any(tag.lower().startswith(nt.lower()) for nt in exclude_tags):
            return []
        sub = ("_" + tag[1:]) if tag.startswith(".") else tag
        # Sanitize characters that are invalid in directory names
        sub = _sanitize_dir_name(sub)
        if not sub:
            continue
        tag_path = os.path.join(base_path, sub)
        os.makedirs(tag_path, exist_ok=True)
        paths.append(os.path.join(tag_path, filename))
    return paths


_RE_INVALID_DIR_CHARS = _re.compile(r'[<>:"|?*\x00-\x1f]')


def _sanitize_dir_name(name: str) -> str:
    """Remove or replace characters that are invalid in directory names."""
    name = _RE_INVALID_DIR_CHARS.sub("_", name)
    # Collapse multiple underscores/spaces
    name = _re.sub(r'_+', '_', name).strip().strip("_")
    return name


# ---------------------------------------------------------------------------
# HTML → Markdown image conversion
# ---------------------------------------------------------------------------

def html_img_to_markdown(text: str) -> str:
    """Replace ``<img src=... alt=...>`` tags with ``![alt](src)`` syntax."""

    def _replace(m: "re.Match") -> str:
        tag = m.group(0)
        src_m = RE_HTML_IMG_SRC.search(tag)
        if not src_m:
            return tag
        src = (src_m.group(2) or "").strip()
        if not src:
            return tag
        alt_m = RE_HTML_IMG_ALT.search(tag)
        alt = (alt_m.group(2) if alt_m else "image").strip() or "image"
        alt = alt.replace("]", r"\]")
        return f"![{alt}]({src})"

    return RE_HTML_IMG_TAG.sub(_replace, text)


# ---------------------------------------------------------------------------
# Reference-style link resolution
# ---------------------------------------------------------------------------

def ref_links_to_inline(text: str) -> str:
    """
    Expand reference-style links and images to inline syntax.

    ``[text][ref]`` → ``[text](url)``
    ``![alt][ref]`` → ``![alt](url)``
    """
    refs = dict(RE_REF_DEF.findall(text))
    if not refs:
        return text

    # Reference images
    text = RE_REF_IMG.sub(
        lambda m: f"![{m.group(1)}]({refs.get(m.group(2), m.group(2))})",
        text,
    )
    text = RE_REF_IMP.sub(
        lambda m: f"![{m.group(1)}]({refs.get(m.group(1), m.group(1))})",
        text,
    )
    # Reference links
    text = RE_REF_LINK.sub(
        lambda m: f"[{m.group(1)}]({refs.get(m.group(2), m.group(2))})",
        text,
    )
    text = RE_REF_LINK_IMP.sub(
        lambda m: (
            f"[{m.group(1)}]({refs[m.group(1)]})"
            if m.group(1) in refs
            else m.group(0)
        ),
        text,
    )
    # Remove reference definitions
    text = RE_REF_CLEAN.sub("", text)
    return text


# ---------------------------------------------------------------------------
# Image link normalisation
# ---------------------------------------------------------------------------

def normalize_local_image_ref(raw_url: Optional[str]) -> str:
    """
    Convert an image URL / path from Markdown or HTML src to a local path token.

    Strips URL encoding, optional ``file://`` prefix, and inline title strings.
    Returns an empty string for ``None`` or empty input.
    """
    if not raw_url:
        return ""
    url = urllib.parse.unquote(str(raw_url)).strip()
    if not url:
        return ""

    # Strip optional inline title: path "title" / path 'title'
    for q in ('"', "'"):
        if q in url:
            head = url[: url.find(q)].rstrip()
            if head:
                url = head
                break

    if url.startswith("<") and url.endswith(">"):
        url = url[1:-1].strip()

    if url.lower().startswith("file://"):
        parsed = urllib.parse.urlparse(url)
        fp = urllib.parse.unquote(parsed.path or "")
        if fp:
            return fp

    return url


# ---------------------------------------------------------------------------
# Misc helpers
# ---------------------------------------------------------------------------

def first_heading(text: str) -> str:
    """Return the first non-empty line stripped of any Markdown heading prefix."""
    for line in text.splitlines():
        line = line.strip()
        if line:
            return RE_MD_HEADING.sub("", line).strip()
    return ""

File: Sources/B2OUCore/Markdown.swift
// Markdown.swift — Pure Markdown transformation functions.
//
// All functions take a string and return a string. No I/O, no config dependencies.

import Foundation

// MARK: - NSRegularExpression Helpers

extension NSRegularExpression {
    func allMatches(in string: String) -> [NSTextCheckingResult] {
        let range = NSRange(string.startIndex..., in: string)
        return matches(in: string, range: range)
    }

    func firstMatch(in string: String) -> NSTextCheckingResult? {
        let range = NSRange(string.startIndex..., in: string)
        return firstMatch(in: string, range: range)
    }

    func replaceAll(in string: String, with template: String) -> String {
        let range = NSRange(string.startIndex..., in: string)
        return stringByReplacingMatches(in: string, range: range, withTemplate: template)
    }

    func replaceAll(in string: String, using block: (NSTextCheckingResult, String) -> String) -> String {
        let nsString = string as NSString
        let range = NSRange(location: 0, length: nsString.length)
        let matches = self.matches(in: string, range: range)

        var result = string
        // Process matches in reverse order to preserve indices
        for match in matches.reversed() {
            guard let matchRange = Range(match.range, in: result) else { continue }
            let replacement = block(match, result)
            result.replaceSubrange(matchRange, with: replacement)
        }
        return result
    }

    func captureGroups(in string: String, match: NSTextCheckingResult) -> [String?] {
        (0..<match.numberOfRanges).map { i in
            let range = match.range(at: i)
            guard range.location != NSNotFound else { return nil }
            return (string as NSString).substring(with: range)
        }
    }
}

// MARK: - Title Sanitization

public func cleanTitle(_ title: String) -> String {
    var result = String(title.prefix(225)).trimmingCharacters(in: .whitespaces)
    if result.isEmpty { result = "Untitled" }
    result = reCleanTitle.replaceAll(in: result, with: "-")
    result = reTrailingDash.replaceAll(in: result, with: "")
    result = result.trimmingCharacters(in: .whitespaces)

    // Ensure UTF-8 byte length fits in 240 bytes
    let maxBytes = 240
    let encoded = Array(result.utf8)
    if encoded.count > maxBytes {
        let truncated = Array(encoded.prefix(maxBytes))
        result = String(bytes: truncated, encoding: .utf8)
            ?? String(decoding: truncated, as: UTF8.self)
        result = result.trimmingCharacters(in: .whitespaces)
    }
    return result
}

// MARK: - Bear → Standard Markdown

public func bearHighlightToMd(_ text: String) -> String {
    reBearHighlight.replaceAll(in: text, with: "==$1==")
}

public func normaliseBearMarkdown(_ text: String) -> String {
    var result = bearHighlightToMd(text)
    result = htmlImgToMarkdown(result)
    result = reBearEmbed.replaceAll(in: result, with: "")
    return result
}

// MARK: - Tag Handling

public func hideTags(_ text: String) -> String {
    reHideTags.replaceAll(in: text, with: "$1")
}

public func extractTags(_ text: String) -> [String] {
    var tags: [String] = []
    for match in reTagPattern1.allMatches(in: text) {
        if let range = Range(match.range(at: 1), in: text) {
            tags.append(String(text[range]))
        }
    }
    for match in reTagPattern2.allMatches(in: text) {
        if let range = Range(match.range(at: 1), in: text) {
            tags.append(String(text[range]))
        }
    }
    return tags
}

public func subPathFromTag(
    basePath: String,
    filename: String,
    text: String,
    makeTagFolders: Bool,
    multiTagFolders: Bool,
    onlyExportTags: [String],
    excludeTags: [String]
) -> [String] {
    let fm = FileManager.default

    if !makeTagFolders {
        if !excludeTags.isEmpty {
            let noteTags = extractTags(text)
            let isExcluded = noteTags.contains { nt in
                excludeTags.contains { et in
                    nt.lowercased().hasPrefix(et.lowercased())
                }
            }
            if isExcluded { return [] }
        }
        return [(basePath as NSString).appendingPathComponent(filename)]
    }

    var tags: [String]
    if multiTagFolders {
        tags = extractTags(text)
        if tags.isEmpty {
            return [(basePath as NSString).appendingPathComponent(filename)]
        }
    } else {
        let t1 = reTagPattern1.firstMatch(in: text)
        let t2 = reTagPattern2.firstMatch(in: text)

        if let t1, let t2 {
            let tag: String
            if t1.range(at: 1).location < t2.range(at: 1).location {
                tag = (text as NSString).substring(with: t1.range(at: 1))
            } else {
                tag = (text as NSString).substring(with: t2.range(at: 1))
            }
            tags = [tag]
        } else if let t1 {
            tags = [(text as NSString).substring(with: t1.range(at: 1))]
        } else if let t2 {
            tags = [(text as NSString).substring(with: t2.range(at: 1))]
        } else {
            return [(basePath as NSString).appendingPathComponent(filename)]
        }
    }

    var paths = [(basePath as NSString).appendingPathComponent(filename)]
    for tag in tags {
        if tag == "/" { continue }
        if !onlyExportTags.isEmpty {
            let match = onlyExportTags.contains { et in
                tag.lowercased().hasPrefix(et.lowercased())
            }
            if !match { continue }
        }
        if excludeTags.contains(where: { tag.lowercased().hasPrefix($0.lowercased()) }) {
            return []
        }
        var sub = tag.hasPrefix(".") ? ("_" + tag.dropFirst()) : tag
        sub = sanitizeDirName(sub)
        if sub.isEmpty { continue }
        let tagPath = (basePath as NSString).appendingPathComponent(sub)
        try? fm.createDirectory(atPath: tagPath, withIntermediateDirectories: true)
        paths.append((tagPath as NSString).appendingPathComponent(filename))
    }
    return paths
}

private func sanitizeDirName(_ name: String) -> String {
    var result = reInvalidDirChars.replaceAll(in: name, with: "_")
    result = reMultipleUnderscores.replaceAll(in: result, with: "_")
    result = result.trimmingCharacters(in: .whitespaces)
    result = result.trimmingCharacters(in: CharacterSet(charactersIn: "_"))
    // Prevent path traversal: strip ".." components
    let components = result.components(separatedBy: "/").filter { $0 != ".." && $0 != "." }
    result = components.joined(separator: "/")
    return result
}

// MARK: - HTML → Markdown Image Conversion

public func htmlImgToMarkdown(_ text: String) -> String {
    reHtmlImgTag.replaceAll(in: text) { match, str in
        let tag = (str as NSString).substring(with: match.range)
        guard let srcMatch = reHtmlImgSrc.firstMatch(in: tag) else { return tag }
        let src = (tag as NSString).substring(with: srcMatch.range(at: 2)).trimmingCharacters(in: .whitespaces)
        if src.isEmpty { return tag }

        var alt = "image"
        if let altMatch = reHtmlImgAlt.firstMatch(in: tag) {
            let altVal = (tag as NSString).substring(with: altMatch.range(at: 2)).trimmingCharacters(in: .whitespaces)
            if !altVal.isEmpty { alt = altVal }
        }
        alt = alt.replacingOccurrences(of: "]", with: "\\]")
        return "![\(alt)](\(src))"
    }
}

// MARK: - Reference-Style Link Resolution

public func refLinksToInline(_ text: String) -> String {
    // Build reference map
    var refs: [String: String] = [:]
    for match in reRefDef.allMatches(in: text) {
        if match.numberOfRanges >= 3 {
            let key = (text as NSString).substring(with: match.range(at: 1))
            let url = (text as NSString).substring(with: match.range(at: 2))
            refs[key] = url
        }
    }
    guard !refs.isEmpty else { return text }

    var result = text

    // Reference images
    result = reRefImg.replaceAll(in: result) { match, str in
        let alt = (str as NSString).substring(with: match.range(at: 1))
        let ref = (str as NSString).substring(with: match.range(at: 2))
        return "![\(alt)](\(refs[ref] ?? ref))"
    }
    result = reRefImp.replaceAll(in: result) { match, str in
        let alt = (str as NSString).substring(with: match.range(at: 1))
        return "![\(alt)](\(refs[alt] ?? alt))"
    }
    // Reference links
    result = reRefLink.replaceAll(in: result) { match, str in
        let label = (str as NSString).substring(with: match.range(at: 1))
        let ref = (str as NSString).substring(with: match.range(at: 2))
        return "[\(label)](\(refs[ref] ?? ref))"
    }
    result = reRefLinkImp.replaceAll(in: result) { match, str in
        let label = (str as NSString).substring(with: match.range(at: 1))
        if let url = refs[label] {
            return "[\(label)](\(url))"
        }
        return (str as NSString).substring(with: match.range)
    }
    // Remove reference definitions
    result = reRefClean.replaceAll(in: result, with: "")
    return result
}

// MARK: - Image Link Normalization

public func normalizeLocalImageRef(_ rawURL: String?) -> String {
    guard var url = rawURL?.trimmingCharacters(in: .whitespaces), !url.isEmpty else { return "" }
    url = url.removingPercentEncoding ?? url

    // Strip optional inline title
    for q in ["\"", "'"] {
        if let idx = url.firstIndex(of: Character(q)) {
            let head = url[url.startIndex..<idx].trimmingCharacters(in: .whitespaces)
            if !head.isEmpty {
                url = head
                break
            }
        }
    }

    if url.hasPrefix("<") && url.hasSuffix(">") {
        url = String(url.dropFirst().dropLast()).trimmingCharacters(in: .whitespaces)
    }

    if url.lowercased().hasPrefix("file://") {
        if let comp = URLComponents(string: url), let path = comp.path.removingPercentEncoding, !path.isEmpty {
            return path
        }
        // Fallback
        let stripped = String(url.dropFirst(7))
        return stripped.removingPercentEncoding ?? stripped
    }

    return url
}

// MARK: - Misc Helpers

public func firstHeading(_ text: String) -> String {
    for line in text.components(separatedBy: .newlines) {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        if !trimmed.isEmpty {
            return reMarkdownHeading.replaceAll(in: trimmed, with: "")
                .trimmingCharacters(in: .whitespaces)
        }
    }
    return ""
}
